### PR TITLE
Rewrite smoketests as python unittests

### PR DESCRIPTION
**Pull request by coolreader18**
# Description of Changes

Title; still need to fix parallel tests

# Expected complexity level and risk

3 - it'll take a bit to get used to the new python format and also it's very easy to conflict with changes made to smoketests while this PR is open, but overall I think it's much preferable to bash.

### CODE REVIEW

**Comment by coolreader18** [comment]
Ok, so - parallel unittest is finicky and probably isn't worth trying to get working right now, BUT I did get it to take only 4 minutes in serial, at least for just the test running.

File: smoketests/__init__.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+import re
+import string
+import random
+import string
+import subprocess
+import json
+import sys
+import shutil
+
+TEST_DIR = Path(__file__).parent
```
Comment (mamcx) [inline smoketests/__init__.py:14]: All path manipulations should use `pathlib` so is possible to run this on Windows.

File: smoketests/__init__.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+def requires_docker(item):
+    if HAVE_DOCKER:
+        return item
+    return unittest.skip("docker not available")(item)
+
+def random_string(k=20):
+    return ''.join(random.choices(string.ascii_letters, k=k))
+
+def extract_fields(cmd_output, field_name):
```
Comment (mamcx) [inline smoketests/__init__.py:50]: I think is fine to assume we have control over the `parsing` input, but maybe add here some docs showing how the data looks like?

File: test/tests/bitcraftmini-pretest.sh [hunk 3]
```diff
@@ -1,19 +0,0 @@
-#!/bin/bash
```
Comment (kazimuth) [inline test/tests/bitcraftmini-pretest.sh:1]: Is this file gonna be replaced with something, or do we not need it anymore?

**Review by kazimuth — COMMENTED** [review]
This looks good in general, there's a lot of changes but it all makes sense. I have a couple of small nits.

Checklist of which tests have been rewritten:
- [x] autoinc*.sh
- [ ] bitcraftmini-pretest.sh
- [x] cancel-reducer.sh
- [x] connect-disconnect-from-cli.sh
- [x] create-project-1.sh
- [x] default-module-clippy.sh
- [x] describe.sh
- [x] detect-wasm-bindgen.sh
- [x] filtering.sh
- [x] identity-*.sh
- [x] module-nested-op.sh
- [x] namespace-custom-cs.sh
- [x] namespace-spacetimedb-cs.sh
- [x] new-user-flow.sh
- [x] panic.sh
- [x] permissions-*.sh
- [x] register-domain.sh (-> domains.py)
- [x] reverse-dns.sh (-> domains.py)
- [x] setname.sh (-> domains.py)
- [x] spacetime-server.sh (-> servers.py)
- [x] update-module.sh (-> modules.py)
- [x] upload-module-*.sh (-> modules.py)
- [x] zz_docker-*.sh

File: smoketests/tests/zz_docker.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    #
+    # Wait until compose believes the health probe succeeds.
+    #
+    # The container may decide to recompile, or grab a coffee at crates.io, or
+    # whatever. In any case, restart doesn't mean the server is up yet.
+    run_cmd("docker", "compose", "up", "--no-recreate", "--detach", "--wait-timeout", "60")
+    # Belts and suspenders!
+    #
+    # The health probe runs inside the container, but that doesn't mean we can
+    # reach it from outside. Ping until we get through.
```
Comment (kazimuth) [inline smoketests/tests/zz_docker.py:20]: ah, docker

File: .test_out
[the remark this answers is not in the document]
Comment (coolreader18) [reply .test_out:1]: oops! i think that's something unittest generates sometimes, I'll add it to gitignore

File: crates/cli/src/subcommands/logs.rs [hunk 5]
```diff
@@ -43,6 +44,13 @@ pub fn cli() -> clap::Command {
                 .help("A flag indicating whether or not to follow the logs")
                 .long_help("A flag that causes logs to not stop when end of the log file is reached, but rather to wait for additional data to be appended to the input."),
         )
+        .arg(
```
[the remark this answers is not in the document]
Comment (coolreader18) [reply crates/cli/src/subcommands/logs.rs:47]: Yep, `Smoktest.logs()` uses it so it doesn't have to do the `cut` weirdness the bash had to do before.

File: test/tests/bitcraftmini-pretest.sh (hunk 3 above)
Comment (coolreader18) [reply test/tests/bitcraftmini-pretest.sh:1]: Yeah, so afaict it's not really needed anymore; it's skipped in CI and I think bitcraft mini isn't even really used for smoketesting stdb anymore.

File: smoketests/unittest_parallel.py [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+
+    process_count = max(0, args.jobs)
+    if process_count == 0:
+        process_count = os.cpu_count()
+
+    # Create the temporary directory (for coverage files)
+    with tempfile.TemporaryDirectory() as temp_dir:
+
+        # Discover tests
+        # with _coverage(args, temp_dir):
```
[the remark this answers is not in the document]
Comment (coolreader18) [reply smoketests/unittest_parallel.py:87]: I kinda wanna leave it in just to make it more clear what's been modified here

File: smoketests/__init__.py (hunk 1 above)
Comment (coolreader18) [reply smoketests/__init__.py:14]: I'm pretty sure I already did that; did you notice somewhere where that's not the case?

Comment (mamcx) [reply smoketests/__init__.py:15]: Like in `SPACETIME_BIN = STDB_DIR / "target/debug/spacetime"`?

(Or this is something new Python covers already)?

**Comment by coolreader18** [comment]
uy, --skip_clippy is abbreviated as `-S` for publish and `-s` for build

File: smoketests/tests/__init__.py
Comment (cloutiertyler) [inline smoketests/tests/__init__.py:1]: What's the point of this file? Am I just out of practice with Python?

**Comment by coolreader18** [comment]
Rebased, but actually made no changes (the last commit ("fix ci") actually does the exact same thing I did in #811 that caused the conflict lol)

**Review by jdetter — APPROVED** [review]
This seems fine, I've gone over some of the tests and everything makes sense here. I've left some questions/comments here but nothing to block the PR from merging 👍

File: smoketests/__main__.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        subprocess.Popen(["docker", "logs", "-f", docker_container])
+        smoketests.HAVE_DOCKER = True
+
+    add_prefix = lambda testlist: [TESTPREFIX + test for test in testlist]
+    import fnmatch
+    excludelist = add_prefix(args.exclude)
+    testlist = add_prefix(args.test)
+
+    loader = ExclusionaryTestLoader(excludelist)
+    loader.testNamePatterns = args.testNamePatterns
```
[the remark this answers is not in the document]
Comment (coolreader18) [reply smoketests/__main__.py:81]: yes!

File: crates/cli/src/tasks/rust.rs [hunk 8]
```diff
@@ -27,7 +27,7 @@ pub(crate) fn build_rust(project_path: &Path, skip_clippy: bool, build_debug: bo
     if !skip_clippy {
         let clippy_conf_dir = tempfile::tempdir()?;
         fs::write(clippy_conf_dir.path().join("clippy.toml"), CLIPPY_TOML)?;
-        println!("checking crate with spacetimedb's clippy configuration");
+        eprintln!("checking crate with spacetimedb's clippy configuration");
```
[the remark this answers is not in the document]
Comment (coolreader18) [reply crates/cli/src/tasks/rust.rs:30]: Not sure if a specific reason (I think maybe it was trying to parse that as json), but at the very least it's the type of message for stderr.

**Pull request merged**